### PR TITLE
Kinetic Accelerator Mods

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
:cl: Kor
add: Kinetic Accelerators are now modular. You can find mod kits in the mining vendor and in RnD.
/:cl:

You have 6 mod slot points on the KA. You can install any combination of mods within the point limit. Remove all mods with a crowbar.

Range Mod*: +1 range, cost of 2

Damage Mod*: +10 damage, cost of 2

Trigger Mod: Golems/Hulks can use it, cost of 2

Indoors Mod: -2 to the divisor used for pressure penalties. Cost of 3. Using all slots on this makes the KA do full damage in doors.

AoE Mod: Old Hyper effect. Cost of 2

Cooldown Mod*: 20% lower cooldown. Cost of 2.

Mods with a \* can be bought from the mining vendor as well as built in RnD. The others are RnD exclusive.

### CODE REVIEW

**Comment by GunHog** [comment]
So how does this work for mining borgies? Can they still receive the hyper KA upgrade?

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm [hunk 1]
```diff
@@ -17,7 +17,7 @@
 	var/unique_frequency = FALSE // modified by KA modkits
 	var/overheat = FALSE
 
-	var/list/max_mod_capacity = 2
+	var/list/max_mod_capacity = 6
```
Comment (RemieRichards) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: WHY

**Comment by TechnoAlchemisto** [comment]
Do non-* mods drop from necropolis crates or what

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm [hunk 2]
```diff
@@ -230,9 +224,90 @@
...
+//Types
+
+/obj/item/modkit/range
+	name = "range increase"
+	desc = "Increases the range of a kinetic accelerator when installed."
+	modifier = 1
+
+/obj/item/modkit/range/modify_projectile(obj/item/projectile/kinetic/K)
+	K.range += modifier
+	return
```
Comment (RemieRichards) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: trailing return, remove it please.

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm [hunk 3]
```diff
@@ -230,9 +224,90 @@
...
+	var/modifier = 1 //For use in any mod kit that has numerical modifiers
 
 /obj/item/modkit/proc/install(obj/item/weapon/gun/energy/kinetic_accelerator/KA)
+	forceMove(KA)
+
+/obj/item/modkit/proc/uninstall(obj/item/weapon/gun/energy/kinetic_accelerator/KA)
+	forceMove(get_turf(KA))
+
+/obj/item/modkit/proc/modify_projectile(obj/item/projectile/kinetic/K)
 	return
```
Comment (RemieRichards) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: trailing return, remove it please.

**Comment by ChangelingRain** [comment]
I suggest making the AoE modkit also do some amount of AoE damage but cost basically all the mod slots.

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm (hunk 1 above)
Comment (KorPhaeron) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: hmmmmmmmm

**Comment by GunHog** [comment]
Please test Mining Cyborg's Hyper KA upgrade, I think the last PR may have rendered them ineffective.

**Comment by KorPhaeron** [comment]
@TechnoAlchemisto 

They come from RnD but SPOOKY MODKITS from chests would be cool

**Comment by KorPhaeron** [comment]
@GunHog 

Borgs should not have been touched

**Comment by GunHog** [comment]
But, the hyper KA was removed, right? The cyborg hyper KA was a subtype of the hyper KA with the two vars set to let borgs use it. It might just be a double child of the normal KA, now.

**Comment by Kromgar** [comment]
Where is the mining mod to make it like the old hyper ka mining style? You said you were going to make it a tradeoff between mining and combat

**Comment by KorPhaeron** [comment]
@Kromgar 

The AoE mod which I forgot to list, but is in the PR

**Comment by kevinz000** [comment]
return HKAs ;_;

**Comment by ChangelingRain** [comment]
Necropolis chests maybe drop a healthsteal mod?
Or a mod that increases cooldown but refunds part of the cooldown on hitting a mob?

**Comment by KorPhaeron** [comment]
@ChangelingRain 

If only miners didn't already have infinite health from legion implants

**Comment by Kromgar** [comment]
Mods that light people on fire. Mods that irradiate people. How aobut making the staff of lava a 6 point mod and shoving that in there. 


@KorPhaeron we can't always get legion souls implanted

**Comment by KorPhaeron** [comment]
Things that do work: Adding the stat altering mods (like range+, and this was the part I was most worried about, because editing projectiles on the fly isn't really supported)

Things that don't work: Crowbarring them out, adding the hypershot mod (just runtimes when you fire it), examining the KA.

Will probably fix tonight or tomorrow

**Comment by KorPhaeron** [comment]
@Kromgar 

I want to add being able to purchase special mods based on what bosses you've killed but I'm not dumb enough to add it to this PR

**Comment by Dorsisdwarf** [comment]
Glad to see the system is still being worked on. Very hype here.

**Comment by lzimann** [comment]
He did it, maybe someone asked this already(tldr), but can you install 3 of the same mod? Like, adding 3 range mods to the ka

**Comment by KorPhaeron** [comment]
Yes

**Comment by KorPhaeron** [comment]
Alright, should be done now

**Comment by GunHog** [comment]
Have you tested that the borg hyper KA upgrade actually gives the borg a working hyper KA?

**Comment by KorPhaeron** [comment]
I hate robots and forgot again

**Comment by KorPhaeron** [comment]
Okay now it's actually done

**Comment by KorPhaeron** [comment]
As a FUN FACT/strange bug I fixed during this: The pressure damage bonus was previously calculated when the shot was loaded, not when you fired. This means if you load a KA outside, you could run back inside and do 40 damage with it.

**Comment by Ergovisavi** [comment]
CDR mod is the optimal choice. OP, please nerf, before merging.

**Comment by Kromgar** [comment]
He did nerf it

**Pull request merged**